src/operator/subgraph_op_common.cc

```diff
@@ -0,0 +1,250 @@
+  imperative::CheckAndInferStorageType(&grad_graph, std::move(dev_masks),
+                                       std::move(storage_type_inputs), true);
+
+  const auto& stypes = grad_graph.GetAttr<StorageTypeVector>("storage_type");
+  DISPATCH_MODE_ASSIGN_CHECK(dispatch_mode, 0, DispatchMode::kFComputeEx);
+  auto &outputs = idx.outputs();
+  CHECK(outputs.size() == out_attrs->size());
+  for (size_t i = 0; i < out_attrs->size(); i++)
+    STORAGE_TYPE_ASSIGN_CHECK(*out_attrs, i, stypes[idx.entry_id(outputs[i])]);
+  return true;
+}
+
+void LoopState::Forward(int iter_no,
+                        std::vector<NDArray> cinputs,
+                        const std::vector<OpReqType>& req,
+                        std::vector<NDArray> coutputs,
+                        bool is_recording) {
+  using namespace nnvm;
+  using namespace imperative;
+
+  bool orig_is_record;
+  if (is_recording)
+    orig_is_record = Imperative::Get()->set_is_recording(true);
+  else
+    orig_is_record = Imperative::Get()->is_recording();
+
+  std::vector<NDArray *> inputs(cinputs.size());
+  std::vector<NDArray *> outputs(coutputs.size());
+  for (size_t i = 0; i < inputs.size(); i++)
+    inputs[i] = &cinputs[i];
+  for (size_t i = 0; i < outputs.size(); i++)
+    outputs[i] = &coutputs[i];
+
+  std::vector<std::pair<std::string, std::string> > kwargs;
+  kwargs.push_back(std::pair<std::string, std::string>("inline_limit", "0"));
+  kwargs.push_back(std::pair<std::string, std::string>("static_alloc", "1"));
+  CachedOpPtr op;
+  if (is_recording && iter_ops.size() > (size_t) iter_no)
+    op = iter_ops[iter_no];
+  else if (!is_recording && iter_ops.size() == 1)
+    op = iter_ops[0];
+
+  // If we need to run backward and we don't have a cached op for this iteration,
+  // we create one for this iteration.
+  if (is_recording && op == nullptr) {
+    op = std::make_shared<CachedOp>(subgraph_sym, kwargs);
+    CHECK_EQ(iter_ops.size(), iter_no);
+    iter_ops.push_back(op);
+  } else if (op == nullptr) {
+    // If we don't need to run backward and this is the first time of
+    // running the iteration, we need to create a new cached op.
+    op = std::make_shared<CachedOp>(subgraph_sym, kwargs);
+    CHECK(iter_ops.empty());
+    iter_ops.push_back(op);
+  }
+  // TODO(zhengda) we need to avoid shape inference and memory plan whenever the op is
+  // called. Currently, CachedOp allocates memory each time Forward is called.
+  // I need to fix this once the PR for static memory allocation in CachedOp is
+  // merged. https://github.com/apache/incubator-mxnet/pull/10817
+  OpStatePtr state = op->Forward(nullptr, inputs, outputs);
+
+  if (is_recording) {
+    all_inputs.push_back(cinputs);
+    all_outputs.push_back(coutputs);
+    all_states.push_back(state);
+  }
+
+  Imperative::Get()->set_is_recording(orig_is_record);
+}
+
+void LoopState::Backward(int iter_no,
+                         std::vector<NDArray> ograds,
```

Comment:
The reason it's not const reference is that we need a copy of the NDArray vector anyway. we need to pass their pointers to cached op, which requests pointers instead of const pointers.